### PR TITLE
[Notifier] Fix encoding of messages with FreeMobileTransport

### PR DESCRIPTION
**Pull request by 94noni**
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix sending accents chars
| License       | MIT

When I first [introduce this transport](https://github.com/symfony/symfony/pull/35690) I used to test it with basic text, now with some French accent I need this fix otherwise the accent chars are not sent

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php [hunk 1]
```diff
@@ -63,7 +63,7 @@ protected function doSend(MessageInterface $message): SentMessage
             'json' => [
                 'user' => $this->login,
                 'pass' => $this->password,
-                'msg' => \utf8_encode($message->getSubject()),
+                'msg' => utf8_encode($message->getSubject()),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: This makes little sense to me: the subject is already UTF-8 encoded, so here we're doubly-encoding the UTF-8 string?
Do you have any link/doc/etc showing that this is correct?

Comment (stof) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: `utf8_encode` is really `convert_iso_1859_1_to_utf8` (there is even php.internals discussion about renaming the function to reflect that). If the input is already UTF-8, the output is corrupted garbage. So this change looks totally wrong to me.

Comment (94noni) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: hi @nicolas-grekas 
Here is a screenshot of my phone message app, first message without and second with this PR code change:

Capture d’écran   2022-01-26 à 15 41 29


my code is like this:

```php
// $texter Symfony/Component/Notifier/TexterInterface
$subject = ''; // this is the text I need with accents
$texter->send(new SmsMessage('06...', $subject));
```

Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: (please send the screenshot again)

Comment (stof) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: looks like you validated the comment form while the upload was still in progress...

Comment (94noni) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: sent again my bad ^^

Comment (stof) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: also, which encoding are you using in your `$subject` variable ? Here, you show an empty string, which won't have encoding issues, so this is clearly not a reproducer.

Comment (94noni) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: @stof I've written for testing `$subject = "àmÏg0";` here for my test before sending this


cf next code reproducer:

```php

// in app code
        $subject = 'àmÏg0';
        $this->texter->send(new SmsMessage('06', $subject));

// in the free mobile transport
        $response = $this->client->request('POST', $endpoint, [
            'json' => [
                'user' => $this->login,
                'pass' => $this->password,
                'msg' => 'normal:' . $message->getSubject(),
            ],
        ]);

        $response = $this->client->request('POST', $endpoint, [
            'json' => [
                'user' => $this->login,
                'pass' => $this->password,
                'msg' => 'utf8_encode:' . utf8_encode($message->getSubject()),
            ],
        ]);

        $response = $this->client->request('POST', $endpoint, [
            'json' => [
                'user' => $this->login,
                'pass' => $this->password,
                'msg' => 'urlencode:' . urlencode($message->getSubject()),
            ],
        ]);
```

Capture d’écran 2022-01-26 à 16 08 20


If needed, I can provide you my {USER} and {PASS} for you to test via mail/slack
The provider doc is here after:

```
L'envoi du SMS se fait en appelant le lien suivant :
https://smsapi.free-mobile.fr/sendmsg
avec les paramètres suivants :
user : votre login
pass : votre clé d'identification générée automatiquement par notre service
msg : le contenu du SMS encodé sous forme d'url (Percent-encoding)
Exemple : Envoyer le message "Hello World !" sur votre mobile : 
https://smsapi.free-mobile.fr/sendmsg?user={USER}&pass={PASS}&msg=Hello%20World%20!

Vous pouvez également, si vous le préférez, envoyer les paramètres en POST.
Dans ce cas, le contenu du message n'a pas besoin d'être encodé.

Le code de retour HTTP indique le succès ou non de l'opération :
200 : Le SMS a été envoyé sur votre mobile.
400 : Un des paramètres obligatoires est manquant.
402 : Trop de SMS ont été envoyés en trop peu de temps.
403 : Le service n'est pas activé sur l'espace abonné, ou login / clé incorrect.
500 : Erreur côté serveur. Veuillez réessayer ultérieurement.
```

Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: can you please try using this?

```php
        $response = $this->client->request('GET', $endpoint, [
            'query' => [
                'user' => $this->login,
                'pass' => $this->password,
                'msg' => 'urlencode:' . rawurlencode($message->getSubject()),
            ],
        ]);
```

Comment (94noni) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: @nicolas-grekas message result is: `urlencode:%C3%A0m%C3%8Fg0` with your code

Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: oh of course, what about this then?

```php
        $response = $this->client->request('GET', $endpoint, [
            'query' => [
                'user' => $this->login,
                'pass' => $this->password,
                'msg' => $message->getSubject(),
            ],
        ]);
```

Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: Also, what about this?
```php
        $response = $this->client->request('POST', $endpoint, [
            'body' => [
                'user' => $this->login,
                'pass' => $this->password,
                'msg' => message->getSubject(),
            ],
        ]);
```

Comment (94noni) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: @nicolas-grekas it works fine result is finally `àmÏg0` !
thus let me swap my PR code and really thank you so much
just can you give the reason behind this? the `json` does specific data alteration that the `query` does not?

Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: I guess the code behind the json-handling logic has been written by someone who messed up with encodings :)

File: src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php [hunk 2]
```diff
@@ -60,7 +60,7 @@ protected function doSend(MessageInterface $message): SentMessage
         $endpoint = sprintf('https://%s', $this->getEndpoint());
 
         $response = $this->client->request('POST', $endpoint, [
-            'json' => [
+            'query' => [
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php:63]: can you please try using `body` here instead?
if we keep `query`, can we keep POST or do we need GET as a method?

File: src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php (hunk 1 above)
Comment (94noni) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php]: let me just try the `body` just to give all results
edit: I got `HTTP/2 400  returned for "https://smsapi.free-mobile.fr/sendmsg".` with `body` and `POST`

File: src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php (hunk 2 above)
Comment (94noni) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php:63]: commented on other discu https://github.com/symfony/symfony/pull/45185#discussion_r792755633
with `query` and `POST` it works well

**Comment by nicolas-grekas** [comment]
Thank you @94noni.

**Pull request merged**